File: plugins/kinetica/skills/kinetica-execute/scripts/modules/image_preview.py
"""PNG-to-ANSI terminal preview module.

Renders a colored ASCII art approximation of a PNG image using the Unicode
half-block technique with 24-bit ANSI colors.  Each terminal cell encodes
2 vertical pixels (foreground = top, background = bottom) using U+2580 (▀).

Zero external dependencies — uses Python built-in zlib and struct.

Exported API:
    render_preview(png_bytes, *, max_width=0, stream=None)
"""

import math
import os
import struct
import sys
import zlib

_PNG_SIGNATURE = b"\x89PNG\r\n\x1a\n"


# ---------------------------------------------------------------------------
# PNG decoder (8-bit RGB / RGBA only)
# ---------------------------------------------------------------------------

def _paeth_predictor(a, b, c):
    """PNG Paeth predictor function."""
    p = a + b - c
    pa = abs(p - a)
    pb = abs(p - b)
    pc = abs(p - c)
    if pa <= pb and pa <= pc:
        return a
    if pb <= pc:
        return b
    return c


def _unfilter_row(filter_type, raw, prev, bpp):
    """Reverse PNG scanline filter for one row.

    Args:
        filter_type: 0-4 (None, Sub, Up, Average, Paeth)
        raw: Raw scanline bytes (without filter byte)
        prev: Previous (already-unfiltered) row, or None for first row
        bpp: Bytes per pixel (3 or 4)

    Returns:
        bytearray of unfiltered scanline
    """
    out = bytearray(raw)
    length = len(out)

    if filter_type == 0:
        return out

    for i in range(length):
        a = out[i - bpp] if i >= bpp else 0
        b = prev[i] if prev else 0

        if filter_type == 1:  # Sub
            out[i] = (out[i] + a) & 0xFF
        elif filter_type == 2:  # Up
            out[i] = (out[i] + b) & 0xFF
        elif filter_type == 3:  # Average
            out[i] = (out[i] + ((a + b) >> 1)) & 0xFF
        elif filter_type == 4:  # Paeth
            c = prev[i - bpp] if (prev and i >= bpp) else 0
            out[i] = (out[i] + _paeth_predictor(a, b, c)) & 0xFF

    return out


def decode_png(data):
    """Decode a PNG buffer into raw pixel data.

    Only supports 8-bit color types 2 (RGB) and 6 (RGBA).

    Args:
        data: Raw PNG file bytes

    Returns:
        dict with keys width, height, bpp, pixels (bytearray), or None
        on unsupported or corrupt input (never raises).
    """
    try:
        if not data or len(data) < 8:
            return None
        if data[:8] != _PNG_SIGNATURE:
            return None

        width = 0
        height = 0
        bit_depth = 0
        color_type = 0
        idat_chunks = []
        pos = 8

        while pos + 8 <= len(data):
            chunk_len = struct.unpack_from(">I", data, pos)[0]
            chunk_type = data[pos + 4:pos + 8]
            data_start = pos + 8

            if chunk_type == b"IHDR":
                width = struct.unpack_from(">I", data, data_start)[0]
                height = struct.unpack_from(">I", data, data_start + 4)[0]
                bit_depth = data[data_start + 8]
                color_type = data[data_start + 9]
            elif chunk_type == b"IDAT":
                idat_chunks.append(data[data_start:data_start + chunk_len])
            elif chunk_type == b"IEND":
                break

            pos = data_start + chunk_len + 4  # +4 for CRC

        if bit_depth != 8:
            return None
        if color_type not in (2, 6):
            return None
        if width == 0 or height == 0:
            return None

        bpp = 3 if color_type == 2 else 4
        compressed = b"".join(idat_chunks)
        raw = zlib.decompress(compressed)

        stride = width * bpp
        pixels = bytearray(height * stride)
        prev_row = None

        for y in range(height):
            row_start = y * (stride + 1)
            filter_type = raw[row_start]
            row_data = raw[row_start + 1:row_start + 1 + stride]
            unfiltered = _unfilter_row(filter_type, row_data, prev_row, bpp)
            pixels[y * stride:(y + 1) * stride] = unfiltered
            prev_row = unfiltered

        return {"width": width, "height": height, "bpp": bpp, "pixels": pixels}
    except Exception:
        return None


# ---------------------------------------------------------------------------
# Scaling
# ---------------------------------------------------------------------------

def scale_to_fit(img_w, img_h, max_cols):
    """Compute output dimensions to fit an image within max_cols terminal columns.

    Does not upscale.  Accounts for 2:1 vertical pixel packing.

    Returns:
        dict with keys cols, rows, scale
    """
    if img_w <= max_cols:
        return {"cols": img_w, "rows": math.ceil(img_h / 2), "scale": 1}
    scale = img_w / max_cols
    cols = max_cols
    rows = math.ceil(img_h / (scale * 2))
    return {"cols": cols, "rows": rows, "scale": scale}


# ---------------------------------------------------------------------------
# Pixel sampling
# ---------------------------------------------------------------------------

def sample_pixel(pixels, img_w, bpp, x, y, img_h):
    """Nearest-neighbor sample a pixel, clamped to image bounds.

    Returns:
        tuple (r, g, b, a)
    """
    px = min(max(round(x), 0), img_w - 1)
    py = min(max(round(y), 0), img_h - 1)
    offset = (py * img_w + px) * bpp
    a = pixels[offset + 3] if bpp == 4 else 255
    return (pixels[offset], pixels[offset + 1], pixels[offset + 2], a)


def area_average(pixels, img_w, bpp, x0, y0, x1, y1, img_h):
    """Area-average (box filter) sample over a rectangle of source pixels.

    Averages all pixels in the integer grid covering [x0, x1) x [y0, y1),
    clamped to image bounds.  Falls back to sample_pixel at midpoint when the
    clamped rectangle is degenerate (zero area).

    Returns:
        tuple (r, g, b, a)
    """
    px_start = max(math.floor(x0), 0)
    py_start = max(math.floor(y0), 0)
    px_end = min(math.ceil(x1), img_w)
    py_end = min(math.ceil(y1), img_h)

    count = (px_end - px_start) * (py_end - py_start)
    if count <= 0:
        mid_x = (x0 + x1) / 2
        mid_y = (y0 + y1) / 2
        return sample_pixel(pixels, img_w, bpp, mid_x, mid_y, img_h)

    sum_r = 0
    sum_g = 0
    sum_b = 0
    sum_a = 0
    for py in range(py_start, py_end):
        row_base = py * img_w * bpp
        for px in range(px_start, px_end):
            offset = row_base + px * bpp
            a = pixels[offset + 3] if bpp == 4 else 255
            sum_r += pixels[offset] * a
            sum_g += pixels[offset + 1] * a
            sum_b += pixels[offset + 2] * a
            sum_a += a

    if sum_a == 0:
        return (0, 0, 0, 0)

    return (
        round(sum_r / sum_a),
        round(sum_g / sum_a),
        round(sum_b / sum_a),
        round(sum_a / count),
    )


# ---------------------------------------------------------------------------
# Alpha blending
# ---------------------------------------------------------------------------

_ALPHA_THRESHOLD = 4


def blend_with_bg(rgba, bg_color):
    """Alpha-over composite a pixel onto a background color.

    Args:
        rgba: (r, g, b, a) pixel
        bg_color: (r, g, b) background

    Returns:
        tuple (r, g, b) blended (fully opaque)
    """
    alpha = rgba[3] / 255
    return (
        round(rgba[0] * alpha + bg_color[0] * (1 - alpha)),
        round(rgba[1] * alpha + bg_color[1] * (1 - alpha)),
        round(rgba[2] * alpha + bg_color[2] * (1 - alpha)),
    )


# ---------------------------------------------------------------------------
# ANSI rendering
# ---------------------------------------------------------------------------

def pixel_to_ansi(top_rgba, bot_rgba, bg_color):
    """Produce an ANSI escape string for one half-block cell.

    Handles transparency: fully transparent pixels yield the terminal's
    native background (no color codes).
    """
    top_transparent = top_rgba[3] <= _ALPHA_THRESHOLD
    bot_transparent = bot_rgba[3] <= _ALPHA_THRESHOLD

    if top_transparent and bot_transparent:
        return " "

    if top_transparent:
        br, bg, bb = blend_with_bg(bot_rgba, bg_color)
        return f"\x1b[38;2;{br};{bg};{bb}m\u2584\x1b[0m"

    if bot_transparent:
        tr, tg, tb = blend_with_bg(top_rgba, bg_color)
        return f"\x1b[38;2;{tr};{tg};{tb}m\u2580\x1b[0m"

    tr, tg, tb = blend_with_bg(top_rgba, bg_color)
    br, bg, bb = blend_with_bg(bot_rgba, bg_color)
    return f"\x1b[38;2;{tr};{tg};{tb}m\x1b[48;2;{br};{bg};{bb}m\u2580\x1b[0m"


# ---------------------------------------------------------------------------
# Orchestrator
# ---------------------------------------------------------------------------

def render_preview(png_bytes, *, max_width=0, stream=None, bg_color=None):
    """Render a PNG buffer as colored ASCII art to a writable stream.

    Args:
        png_bytes: Raw PNG file bytes
        max_width: Max columns (0 = auto-detect from terminal)
        stream: Writable file object (default: sys.stderr)
        bg_color: (r, g, b) background for alpha blending (default: (0, 0, 0))
    """
    if bg_color is None:
        bg_color = (0, 0, 0)
    if stream is None:
        stream = sys.stderr

    # Skip when output is piped (not a terminal)
    if not hasattr(stream, "isatty") or not stream.isatty():
        return

    try:
        img = decode_png(png_bytes)
        if not img:
            stream.write("Warning: could not decode PNG for preview\n")
            return

        if max_width <= 0:
            try:
                max_width = os.get_terminal_size().columns
            except (OSError, ValueError):
                max_width = 80

        dims = scale_to_fit(img["width"], img["height"], max_width)
        cols = dims["cols"]
        rows = dims["rows"]
        scale = dims["scale"]

        lines = []
        for row in range(rows):
            line_parts = []
            for col in range(cols):
                x0 = col * scale
                x1 = (col + 1) * scale
                top_y0 = row * 2 * scale
                top_y1 = (row * 2 + 1) * scale
                bot_y0 = top_y1
                bot_y1 = (row * 2 + 2) * scale

                if scale > 1:
                    top_rgba = area_average(
                        img["pixels"], img["width"], img["bpp"],
                        x0, top_y0, x1, top_y1, img["height"],
                    )
                    bot_rgba = area_average(
                        img["pixels"], img["width"], img["bpp"],
                        x0, bot_y0, x1, bot_y1, img["height"],
                    )
                else:
                    top_rgba = sample_pixel(
                        img["pixels"], img["width"], img["bpp"],
                        x0, top_y0, img["height"],
                    )
                    bot_rgba = sample_pixel(
                        img["pixels"], img["width"], img["bpp"],
                        x0, bot_y0, img["height"],
                    )
                line_parts.append(pixel_to_ansi(top_rgba, bot_rgba, bg_color))
            lines.append("".join(line_parts))
        stream.write("\n".join(lines) + "\n")
    except Exception:
        stream.write("Warning: preview rendering failed\n")

File: plugins/kinetica/skills/kinetica-execute/scripts/modules/image-preview.js
'use strict';

/**
 * PNG-to-ANSI terminal preview module.
 *
 * Renders a colored ASCII art approximation of a PNG image using the Unicode
 * half-block technique with 24-bit ANSI colors.  Each terminal cell encodes
 * 2 vertical pixels (foreground = top, background = bottom) using U+2580 (▀).
 *
 * Zero external dependencies — uses Node.js built-in zlib for PNG inflate.
 *
 * Exported API:
 *   renderPreview(pngBuffer, opts)   — write ASCII art to a stream
 *   decodePng(buffer)                — parse PNG → pixel array (exported for testing)
 *   scaleToFit(imgW, imgH, maxCols) — compute scaled dimensions
 *   samplePixel(pixels, imgW, bpp, x, y, imgH) — nearest-neighbor sample → [r,g,b,a]
 *   areaAverage(pixels, imgW, bpp, x0, y0, x1, y1, imgH) — box-filter sample → [r,g,b,a]
 *   blendWithBg(rgba, bgColor)      — alpha-over composite onto background
 *   pixelToAnsi(topRgba, botRgba, bgColor) — ANSI escape for one half-block cell
 */

const zlib = require('zlib');

// ---------------------------------------------------------------------------
// PNG decoder (8-bit RGB / RGBA only)
// ---------------------------------------------------------------------------

const PNG_SIGNATURE = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);

/**
 * Reverse PNG scanline filter for one row.
 * @param {number} filterType  0-4 (None, Sub, Up, Average, Paeth)
 * @param {Buffer} raw         Raw scanline bytes (without filter byte)
 * @param {Buffer|null} prev   Previous (already-unfiltered) scanline, or null for first row
 * @param {number} bpp         Bytes per pixel (3 or 4)
 * @returns {Buffer} Unfiltered scanline
 */
function unfilterRow(filterType, raw, prev, bpp) {
  const out = Buffer.from(raw);
  const len = out.length;

  if (filterType === 0) return out; // None

  for (let i = 0; i < len; i++) {
    const a = i >= bpp ? out[i - bpp] : 0;
    const b = prev ? prev[i] : 0;

    if (filterType === 1) {
      // Sub
      out[i] = (out[i] + a) & 0xff;
    } else if (filterType === 2) {
      // Up
      out[i] = (out[i] + b) & 0xff;
    } else if (filterType === 3) {
      // Average
      out[i] = (out[i] + ((a + b) >>> 1)) & 0xff;
    } else if (filterType === 4) {
      // Paeth
      const c = (prev && i >= bpp) ? prev[i - bpp] : 0;
      out[i] = (out[i] + paethPredictor(a, b, c)) & 0xff;
    }
  }
  return out;
}

function paethPredictor(a, b, c) {
  const p = a + b - c;
  const pa = Math.abs(p - a);
  const pb = Math.abs(p - b);
  const pc = Math.abs(p - c);
  if (pa <= pb && pa <= pc) return a;
  if (pb <= pc) return b;
  return c;
}

/**
 * Decode a PNG buffer into raw pixel data.
 * Only supports 8-bit color types 2 (RGB) and 6 (RGBA).
 *
 * @param {Buffer} buffer  Raw PNG file bytes
 * @returns {{ width: number, height: number, bpp: number, pixels: Buffer } | null}
 *   null on unsupported or corrupt input (never throws)
 */
function decodePng(buffer) {
  try {
    if (!buffer || buffer.length < 8) return null;
    if (buffer.compare(PNG_SIGNATURE, 0, 8, 0, 8) !== 0) return null;

    let width = 0;
    let height = 0;
    let bitDepth = 0;
    let colorType = 0;
    const idatChunks = [];
    let pos = 8;

    while (pos + 8 <= buffer.length) {
      const chunkLen = buffer.readUInt32BE(pos);
      const chunkType = buffer.toString('ascii', pos + 4, pos + 8);
      const dataStart = pos + 8;

      if (chunkType === 'IHDR') {
        width = buffer.readUInt32BE(dataStart);
        height = buffer.readUInt32BE(dataStart + 4);
        bitDepth = buffer[dataStart + 8];
        colorType = buffer[dataStart + 9];
      } else if (chunkType === 'IDAT') {
        idatChunks.push(buffer.subarray(dataStart, dataStart + chunkLen));
      } else if (chunkType === 'IEND') {
        break;
      }

      pos = dataStart + chunkLen + 4; // +4 for CRC
    }

    if (bitDepth !== 8) return null;
    if (colorType !== 2 && colorType !== 6) return null;
    if (width === 0 || height === 0) return null;

    const bpp = colorType === 2 ? 3 : 4;
    const compressed = Buffer.concat(idatChunks);
    const raw = zlib.inflateSync(compressed);

    const stride = width * bpp;
    const pixels = Buffer.alloc(height * stride);
    let prevRow = null;

    for (let y = 0; y < height; y++) {
      const rowStart = y * (stride + 1); // +1 for filter byte
      const filterType = raw[rowStart];
      const rowData = raw.subarray(rowStart + 1, rowStart + 1 + stride);
      const unfiltered = unfilterRow(filterType, rowData, prevRow, bpp);
      unfiltered.copy(pixels, y * stride);
      prevRow = unfiltered;
    }

    return { width, height, bpp, pixels };
  } catch (_err) {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Scaling
// ---------------------------------------------------------------------------

/**
 * Compute output dimensions to fit an image within maxCols terminal columns.
 * Does not upscale.  Accounts for 2:1 vertical pixel packing.
 *
 * @param {number} imgW  Source image width
 * @param {number} imgH  Source image height
 * @param {number} maxCols  Maximum terminal columns
 * @returns {{ cols: number, rows: number, scale: number }}
 */
function scaleToFit(imgW, imgH, maxCols) {
  if (imgW <= maxCols) {
    // No downscale needed — 1 pixel per column
    return { cols: imgW, rows: Math.ceil(imgH / 2), scale: 1 };
  }
  const scale = imgW / maxCols;
  const cols = maxCols;
  const rows = Math.ceil(imgH / (scale * 2));
  return { cols, rows, scale };
}

// ---------------------------------------------------------------------------
// Pixel sampling
// ---------------------------------------------------------------------------

/**
 * Nearest-neighbor sample a pixel, clamped to image bounds.
 *
 * @param {Buffer} pixels  Raw pixel buffer
 * @param {number} imgW    Image width
 * @param {number} bpp     Bytes per pixel (3 or 4)
 * @param {number} x       Pixel X coordinate (may be fractional)
 * @param {number} y       Pixel Y coordinate (may be fractional)
 * @param {number} imgH    Image height
 * @returns {number[]}     [r, g, b, a]
 */
function samplePixel(pixels, imgW, bpp, x, y, imgH) {
  const px = Math.min(Math.max(Math.round(x), 0), imgW - 1);
  const py = Math.min(Math.max(Math.round(y), 0), imgH - 1);
  const offset = (py * imgW + px) * bpp;
  const a = bpp === 4 ? pixels[offset + 3] : 255;
  return [pixels[offset], pixels[offset + 1], pixels[offset + 2], a];
}

/**
 * Area-average (box filter) sample over a rectangle of source pixels.
 * Averages all pixels in the integer grid covering [x0, x1) × [y0, y1),
 * clamped to image bounds.  Falls back to samplePixel at midpoint when the
 * clamped rectangle is degenerate (zero area).
 *
 * @param {Buffer} pixels  Raw pixel buffer
 * @param {number} imgW    Image width
 * @param {number} bpp     Bytes per pixel (3 or 4)
 * @param {number} x0      Left edge (inclusive, may be fractional)
 * @param {number} y0      Top edge (inclusive, may be fractional)
 * @param {number} x1      Right edge (exclusive, may be fractional)
 * @param {number} y1      Bottom edge (exclusive, may be fractional)
 * @param {number} imgH    Image height
 * @returns {number[]}     [r, g, b, a]
 */
function areaAverage(pixels, imgW, bpp, x0, y0, x1, y1, imgH) {
  const pxStart = Math.max(Math.floor(x0), 0);
  const pyStart = Math.max(Math.floor(y0), 0);
  const pxEnd = Math.min(Math.ceil(x1), imgW);
  const pyEnd = Math.min(Math.ceil(y1), imgH);

  const count = (pxEnd - pxStart) * (pyEnd - pyStart);
  if (count <= 0) {
    const midX = (x0 + x1) / 2;
    const midY = (y0 + y1) / 2;
    return samplePixel(pixels, imgW, bpp, midX, midY, imgH);
  }

  let sumR = 0;
  let sumG = 0;
  let sumB = 0;
  let sumA = 0;
  for (let py = pyStart; py < pyEnd; py++) {
    const rowBase = py * imgW * bpp;
    for (let px = pxStart; px < pxEnd; px++) {
      const offset = rowBase + px * bpp;
      const a = bpp === 4 ? pixels[offset + 3] : 255;
      sumR += pixels[offset] * a;
      sumG += pixels[offset + 1] * a;
      sumB += pixels[offset + 2] * a;
      sumA += a;
    }
  }

  if (sumA === 0) {
    return [0, 0, 0, 0];
  }

  return [
    Math.round(sumR / sumA),
    Math.round(sumG / sumA),
    Math.round(sumB / sumA),
    Math.round(sumA / count),
  ];
}

// ---------------------------------------------------------------------------
// Alpha blending
// ---------------------------------------------------------------------------

const ALPHA_THRESHOLD = 4;

/**
 * Alpha-over composite a pixel onto a background color.
 *
 * @param {number[]} rgba     [r, g, b, a]
 * @param {number[]} bgColor  [r, g, b] background
 * @returns {number[]}        [r, g, b] blended (fully opaque)
 */
function blendWithBg(rgba, bgColor) {
  const alpha = rgba[3] / 255;
  return [
    Math.round(rgba[0] * alpha + bgColor[0] * (1 - alpha)),
    Math.round(rgba[1] * alpha + bgColor[1] * (1 - alpha)),
    Math.round(rgba[2] * alpha + bgColor[2] * (1 - alpha)),
  ];
}

// ---------------------------------------------------------------------------
// ANSI rendering
// ---------------------------------------------------------------------------

/**
 * Produce an ANSI escape string for one half-block cell.
 * Handles transparency: fully transparent pixels yield the terminal's
 * native background (no color codes).
 *
 * @param {number[]} topRgba  [r, g, b, a] for top pixel
 * @param {number[]} botRgba  [r, g, b, a] for bottom pixel
 * @param {number[]} bgColor  [r, g, b] background for blending
 * @returns {string}
 */
function pixelToAnsi(topRgba, botRgba, bgColor) {
  const topTransparent = topRgba[3] <= ALPHA_THRESHOLD;
  const botTransparent = botRgba[3] <= ALPHA_THRESHOLD;

  if (topTransparent && botTransparent) {
    return ' ';
  }

  if (topTransparent) {
    // Only bottom pixel visible — lower half-block as foreground
    const [br, bg, bb] = blendWithBg(botRgba, bgColor);
    return `\x1b[38;2;${br};${bg};${bb}m\u2584\x1b[0m`;
  }

  if (botTransparent) {
    // Only top pixel visible — upper half-block as foreground
    const [tr, tg, tb] = blendWithBg(topRgba, bgColor);
    return `\x1b[38;2;${tr};${tg};${tb}m\u2580\x1b[0m`;
  }

  // Both opaque — upper half-block with fg=top, bg=bottom
  const [tr, tg, tb] = blendWithBg(topRgba, bgColor);
  const [br, bg, bb] = blendWithBg(botRgba, bgColor);
  return `\x1b[38;2;${tr};${tg};${tb}m\x1b[48;2;${br};${bg};${bb}m\u2580\x1b[0m`;
}

// ---------------------------------------------------------------------------
// Orchestrator
// ---------------------------------------------------------------------------

/**
 * Render a PNG buffer as colored ASCII art to a writable stream.
 *
 * @param {Buffer} pngBuffer  Raw PNG file bytes
 * @param {object} [opts]     Options
 * @param {number} [opts.maxWidth=0]  Max columns (0 = auto-detect from stream)
 * @param {object} [opts.stream]      Writable stream (default: process.stderr)
 * @param {number[]} [opts.bgColor=[0,0,0]]  Background color for alpha blending
 */
function renderPreview(pngBuffer, opts = {}) {
  const stream = opts.stream || process.stderr;
  const bgColor = opts.bgColor || [0, 0, 0];

  // Skip when output is piped (not a terminal)
  if (!stream.isTTY) return;

  try {
    const img = decodePng(pngBuffer);
    if (!img) {
      stream.write('Warning: could not decode PNG for preview\n');
      return;
    }

    const maxCols = opts.maxWidth || stream.columns || 80;
    const { cols, rows, scale } = scaleToFit(img.width, img.height, maxCols);

    const lines = [];
    for (let row = 0; row < rows; row++) {
      let line = '';
      for (let col = 0; col < cols; col++) {
        const x0 = col * scale;
        const x1 = (col + 1) * scale;
        const topY0 = row * 2 * scale;
        const topY1 = (row * 2 + 1) * scale;
        const botY0 = topY1;
        const botY1 = (row * 2 + 2) * scale;

        const topRgba = scale > 1
          ? areaAverage(img.pixels, img.width, img.bpp, x0, topY0, x1, topY1, img.height)
          : samplePixel(img.pixels, img.width, img.bpp, x0, topY0, img.height);
        const botRgba = scale > 1
          ? areaAverage(img.pixels, img.width, img.bpp, x0, botY0, x1, botY1, img.height)
          : samplePixel(img.pixels, img.width, img.bpp, x0, botY0, img.height);
        line += pixelToAnsi(topRgba, botRgba, bgColor);
      }
      lines.push(line);
    }
    stream.write(lines.join('\n') + '\n');
  } catch (_err) {
    stream.write('Warning: preview rendering failed\n');
  }
}

module.exports = { renderPreview, decodePng, scaleToFit, samplePixel, areaAverage, blendWithBg, pixelToAnsi };
